### PR TITLE
docs: Explain exit code 2

### PR DESCRIPTION
**Pull request by holmanb**
## Proposed Commit Message
```
doc: Explain breaking change in status code
```

## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

File: doc/rtd/howto/return_codes.rst
Comment (TheRealFalcon) [inline doc/rtd/howto/return_codes.rst:43]: This sentence feels a bit long. Maybe convert to bullets or somehow separate into two sentences.

File: doc/rtd/howto/return_codes.rst
Comment (TheRealFalcon) [inline doc/rtd/howto/return_codes.rst:67]:
```suggestion
autocompletion. Several different text editors are capable of this.
```

File: doc/rtd/howto/return_codes.rst
Comment (TheRealFalcon) [inline doc/rtd/howto/return_codes.rst:70]: Install this extension: https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml . Create a file called `cloud-config.yaml` and it should work automatically.

Also see test steps here: https://github.com/canonical/cloud-init/pull/1479

File: doc/rtd/howto/return_codes.rst
Comment (TheRealFalcon) [inline doc/rtd/howto/return_codes.rst:75]: Should link to https://cloudinit.readthedocs.io/en/latest/howto/debug_user_data.html

File: doc/rtd/howto/return_codes.rst
Comment (TheRealFalcon) [inline doc/rtd/howto/return_codes.rst:83]: Should link to https://cloudinit.readthedocs.io/en/latest/howto/status.html#cloud-init-s-extended-status

File: doc/rtd/howto/return_codes.rst
Comment (TheRealFalcon) [inline doc/rtd/howto/return_codes.rst:87]: I think this section can be merged with the one above it

File: doc/rtd/howto/return_codes.rst
Comment (TheRealFalcon) [inline doc/rtd/howto/return_codes.rst:109]: Since we could neglect to update this for a while, let's not mention recently and instead just say something like "As of 23.4...".

Also, I don't remember the exact page, but I remember you expressing frustration about an older page talking about this history of how things were years ago instead of describing how it works. I could see that quickly being relevant here and I'm not sure the two paragraphs above this one are necessary.

File: doc/rtd/howto/return_codes.rst
Comment (TheRealFalcon) [inline doc/rtd/howto/return_codes.rst:119]: Shell might be a better example here just as that's the general the lowest common denominator for scripting.

File: doc/rtd/howto/return_codes.rst
Comment (TheRealFalcon) [inline doc/rtd/howto/return_codes.rst:163]: Not sure I'd call out specific distros here.

Also, this paragraph feels a little weird to call out here. We don't generally point out every possible breaking change that has been patched out in various downstreams. Not sure we need anything, but if so I'd reduce it to a note with something along the lines of "Distributions have likely removed this behavior for stable releases".

File: doc/rtd/howto/return_codes.rst
Comment (TheRealFalcon) [inline doc/rtd/howto/return_codes.rst:12]:
```suggestion
Background
```
Only because history is boring :smile:

File: doc/rtd/howto/return_codes.rst
Comment (TheRealFalcon) [inline doc/rtd/howto/return_codes.rst:18]: I don't think we need the sentences before this. I don't care about the early days of cloud-init, more just the reasoning behind why things work the way they do.

**Review by TheRealFalcon — CHANGES_REQUESTED** [review]
I left some comments inline. Docs don't build on this branch and after fixing one issue and running into another, I'll leave it to you to fix that. 

More generally, the page feels very disjoint. Much of the history and description of various things we've added to report warnings aren't super relevant to the question being asked in the heading. I think it is useful information though, and I'm wondering if this would make more sense as a "Dealing with problems" explanation. The `why exit 2` can just be a subset of a larger explanation.

We should probably also update https://cloudinit.readthedocs.io/en/latest/reference/cli.html#status

File: doc/rtd/howto/return_codes.rst
Comment (holmanb) [reply doc/rtd/howto/return_codes.rst:119]: I was originally going to use shell, but it doesn't lend well to making use of the machine-readable output available in cloud-init, which real languages would want to make use of, which is why I went with Python.

File: doc/rtd/howto/return_codes.rst
Comment (holmanb) [reply doc/rtd/howto/return_codes.rst]: Thanks!

**Comment by holmanb** [comment]
@TheRealFalcon Ready for re-review, I think I've addressed everything except the preference to shell. Shell isn't great for handling machine-readable output without jq, and I think we're in agreement that avoiding that if possible. Python is pretty readable for most people, but I can add another language example if that works (Go / Rust / ???). If you feel strongly about using shell I'll do that, but my goal with those examples was to demonstrate the proper way of handling `exit 2`, which in my opinion is logging the errors raised from cloud-init at a minimum.

**Comment by holmanb** [comment]
> @TheRealFalcon Ready for re-review, I think I've addressed everything except the preference to shell. Shell isn't great for handling machine-readable output without jq, and I think we're in agreement that avoiding that if possible. Python is pretty readable for most people, but I can add another language example if that works (Go / Rust / ???). If you feel strongly about using shell I'll do that, but my goal with those examples was to demonstrate the proper way of handling `exit 2`, which in my opinion is logging the errors raised from cloud-init at a minimum.

I suppose I could use an example that just prints _all_ of the output from `cloud-init status --long` - thoughts?

**Review by TheRealFalcon — APPROVED** [review]
LGTM! Should we get a review from @s-makin before merging?

File: doc/rtd/explanation/return_codes.rst
Comment (s-makin) [inline doc/rtd/explanation/return_codes.rst:64]:
```suggestion
used for YAML syntax checking, warnings when invalid keys are used, and
```

File: doc/rtd/explanation/return_codes.rst
Comment (s-makin) [inline doc/rtd/explanation/return_codes.rst:67]:
```suggestion
can install the `extension`_ and then a file named ``cloud-config.yaml``
```

File: doc/rtd/explanation/return_codes.rst
Comment (s-makin) [inline doc/rtd/explanation/return_codes.rst:68]:
```suggestion
will automatically use cloud-init's JSON schema.
```

File: doc/rtd/explanation/return_codes.rst
Comment (s-makin) [inline doc/rtd/explanation/return_codes.rst:83]:
```suggestion
Cloud-init historically used two return codes from the :code:`cloud-init status`
```

File: doc/rtd/reference/cli.rst
Comment (s-makin) [inline doc/rtd/reference/cli.rst:407]:
```suggestion
Report cloud-init's current status.
```

**Comment by s-makin** [comment]
> You made some good points about the Python. Cloud-init is in python, so the python syntax shouldn't cause anybody reading the docs any problems.

My $0.02 is basically this - it's a Python project, having the Python code in the docs should be totally fine and expected imo. I don't see anything that's in there with zero context so given that all the code is explained, I'd be happy to see it as-is.

**Review by s-makin — APPROVED** [review]
Thanks @holmanb :) lgtm!

**Comment by holmanb** [comment]
Thanks for the reviews @TheRealFalcon @s-makin!

**Pull request merged**
